### PR TITLE
Sparse Integration

### PR DESCRIPTION
**Pull request by KSGulin**
Added:
- Export functionality
- DDP support
- Updated dependencies

### CODE REVIEW

File: export.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    batch_shape = (args.batch_size, *args.image_shape)
+    state_dict = torch.load(args.checkpoint)
+    model.load_state_dict(state_dict['state_dict'])
+
+    # export to onnx graph   
+    export_onnx(
+        module=model,
+        sample_batch=torch.randn(*batch_shape),
+        file_path=str(args.name),
+        convert_qat=True,
```
Comment (rahul-tuli) [inline export.py]: `convert_qat` is always `True`, use the negation of `no_qat` argument maybe?

File: export.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+@dataclass
+class ExportArgs:
+    """
+    Typed arguments for exporting a ViT model to ONNX
+    """
+
+    checkpoint: Path
+    config: str
+    recipe: str
+    no_qat: bool
```
Comment (rahul-tuli) [inline export.py]: maybe rename to `no_qat_conv`

File: export.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            drop_rate=cfg.drop,
+            drop_connect_rate=cfg.drop_connect, 
+            drop_path_rate=cfg.drop_path,
+            drop_block_rate=cfg.drop_block,
+            global_pool=cfg.gp,
+            bn_tf=cfg.bn_tf,
+            bn_momentum=cfg.bn_momentum,
+            bn_eps=cfg.bn_eps,
+            scriptable=cfg.torchscript,
+            checkpoint_path=None)
```
Comment (rahul-tuli) [inline export.py]: nit: closing brace on newline? Comma after checkpoint path? Something like the following:
```suggestion
            checkpoint_path=None,
            )
```

File: train.py [hunk 4]
```diff
@@ -266,6 +266,8 @@
                     help='how many training processes to use (default: 1)')
 parser.add_argument('--save-images', action='store_true', default=False,
                     help='save images of input bathes every log interval for debugging')
+parser.add_argument('--distributed', action='store_true', default=False,
```
Comment (rahul-tuli) [inline train.py:269]: nit: Can remove `default=False` if `action='store_true'` used

File: export.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                        The image shape in (C, S, S) format to use for exporting
+                        the Model graph to ONNX; Defaults to (3, 550, 550)
+  --save-dir SAVE_DIR, -s SAVE_DIR
+                        The directory to save exported models to; Defaults to
+                        "./exported_models"
+  --name NAME, -n NAME  The name to use for saving the exported ONNX model
+##########
+Example usage:
+python export.py --checkpoint ./checkpoints/vit_base_patch32_224-224_pruned.pth.tar \
+    --recipe ./recipes/vit_base.85.recal.config.yaml \
```
Comment (rahul-tuli) [inline export.py]: Typo: Remove trailing backslash

**Review by rahul-tuli — CHANGES_REQUESTED** [review]
Looks good overall, needs a few small fixes. Good Work!

File: export.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        while self.name.exists():
+            existence_counter += 1
+            self.name = self.name.parent / f"{filename}-{existence_counter}.onnx"
+        return self.name
+
+
+def parse_args() -> ExportArgs:
+    """
+    Add and parse arguments for exporting a ViT model to ONNX
+    :return: A ExportArgs Object with typed arguments
```
Comment (rahul-tuli) [inline export.py]: Sphinx needs an empty line before params to generate docs
```suggestion
    Add and parse arguments for exporting a ViT model to ONNX
    
    :return: A ExportArgs Object with typed arguments
```

File: train.py (hunk 4 above)
Comment (KSGulin) [reply train.py:269]: Keeping default=False seems to adhere to the TIMM conventions. Not sure if convention is a relic or used to make the defaults more explicit to other devs.

**Comment by KSGulin** [comment]
Thanks for catching those @rahul-tuli! Addressed all but the one I commented on

File: export.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+  --checkpoint CHECKPOINT
+                        The ViT pytorch checkpoint to export
+  --config CONFIG, -c CONFIG
+                        The config used to train the ViT model,
+                        for ex: Defaults to look for args.yaml in 
+                        checkpoint directory.
+  --recipe RECIPE, -r RECIPE
+                        Path or SparseZoo stub to the recipe used for training,
+                        omit if no recipe used.
+  --no-qat, -N          Flag to prevent conversion of a QAT(Quantization Aware
```
Comment (rahul-tuli) [inline export.py]: Should this be `--no-qat-conv`?

**Review by rahul-tuli — CHANGES_REQUESTED** [review]

File: export.py (hunk 7 above)
Comment (KSGulin) [reply export.py]: Good catch! Fixed.

**Review by spacemanidol — DISMISSED** [review]

File: export.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+  --checkpoint CHECKPOINT
+                        The ViT pytorch checkpoint to export
+  --config CONFIG, -c CONFIG
+                        The config used to train the ViT model,
+                        for ex: Defaults to look for args.yaml in 
+                        checkpoint directory.
+  --recipe RECIPE, -r RECIPE
+                        Path or SparseZoo stub to the recipe used for training,
+                        omit if no recipe used.
+  --no-qat-conv, -N          Flag to prevent conversion of a QAT(Quantization Aware
```
Comment (rahul-tuli) [inline export.py]: nit: remove extra space

File: requirements.txt [hunk 9]
```diff
@@ -1,3 +1,4 @@
-torch>=1.4.0
-torchvision>=0.5.0
+torch>=1.4.0,<1.10
+torchvision>=0.5.0,<0.11
 pyyaml
+wandb
```
Comment (rahul-tuli) [inline requirements.txt]: `wandb` usage is optional so let's not put it in `requirements.txt`

**Review by rahul-tuli — DISMISSED** [review]
Minor nits, but looks good overall

**Comment by KSGulin** [comment]
Addressed, thanks Rahul!

**Review by rahul-tuli — DISMISSED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**